stl/inc/ranges

```diff
@@ -6837,6 +6837,602 @@ namespace ranges {
+    private:
+        friend class zip_view<_ViewTypes...>;
+
+        using _Iterator_tuple_type = tuple<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>...>;
+
+        constexpr explicit _Iterator(_Iterator_tuple_type _Current) noexcept(
+            is_nothrow_move_constructible_v<_Iterator_tuple_type>)
+            : _M_Current(_STD move(_Current)) {}
+
+        static constexpr bool _Is_nothrow_dereferenceable =
+            (noexcept(*(_STD declval<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>())) && ...);
+
+        static constexpr bool _Is_nothrow_preincrementable =
+            (noexcept(++(_STD declval<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>())) && ...);
+
+        static constexpr bool _Is_nothrow_predecrementable =
+            (noexcept(--(_STD declval<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>())) && ...);
+
+    public:
+        // NOTE: iterator_category is conditionally present based on whether or not it exists as
+        // a type alias in _Iterator_category_container.
+
+        using iterator_concept = conditional_t<_Zip::_All_random_access<_IsConst, _ViewTypes...>,
+            random_access_iterator_tag,
+            conditional_t<_Zip::_All_bidirectional<_IsConst, _ViewTypes...>, bidirectional_iterator_tag,
+                conditional_t<_Zip::_All_forward<_IsConst, _ViewTypes...>, forward_iterator_tag, input_iterator_tag>>>;
+
+        using value_type      = tuple<range_value_t<_Maybe_const<_IsConst, _ViewTypes>>...>;
+        using difference_type = common_type_t<range_difference_t<_Maybe_const<_IsConst, _ViewTypes>>...>;
+
+        _Iterator() noexcept(is_nothrow_default_constructible_v<_Iterator_tuple_type>) = default;
+
+        constexpr _Iterator(_Iterator<!_IsConst> _Rhs) noexcept(
+            (is_nothrow_move_constructible_v<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>> && ...)) // strengthened
+            requires(_IsConst
+                     && (convertible_to<iterator_t<_ViewTypes>, iterator_t<_Maybe_const<_IsConst, _ViewTypes>>> && ...))
+            : _M_Current(_STD move(_Rhs._M_Current)) {}
+
+        _NODISCARD constexpr auto operator*() const noexcept(_Is_nothrow_dereferenceable) /* strengthened */ {
+            return _Zip::_Tuple_transform([](auto& _Itr) -> decltype(auto) { return *_Itr; }, _M_Current);
+        }
+
+        constexpr _Iterator& operator++() noexcept(_Is_nothrow_preincrementable) /* strengthened */ {
+            _Zip::_Tuple_for_each([](auto& _Itr) { ++_Itr; }, _M_Current);
+            return *this;
+        }
+
+        constexpr void operator++(int) noexcept(_Is_nothrow_preincrementable) /* strengthened */ {
+            ++(*this);
+        }
+
+        constexpr _Iterator operator++(int) noexcept(
+            _Is_nothrow_preincrementable&& is_nothrow_copy_assignable_v<_Iterator<_IsConst>>) // strengthened
+            requires _Zip::_All_forward<_IsConst, _ViewTypes...> {
+            const auto _Temp = *this;

```

Comment:
Question for other maintainers and longtime contributors: I forget, do we need to say `auto` here instead of `const auto` to be maximally friendly to guaranteed copy elision/deferred temporary materialization/the Named Return Value Optimization? (As a special exception to our usual preference of using `const` whenever possible.) Occurs below.

Reply:
**tl;dr:** Just keep using `const`.

I've done a bit of research on this topic and I found the following on cppreference (applies to C++17 and later, emphasis mine):

> Under the following circumstances, the compilers are required to omit the copy and move construction of class objects, even if the copy/move constructor and the destructor have observable side-effects.
>
>[ ... ]
>
>* In the initialization of an object, when the initializer expression is a prvalue of the same class type (**ignoring cv-qualification**) as the variable type:
>[...]

So, according to the standard, const qualifications have no effect on copy elision, which is a mandatory feature of C++ compilers. Thus, we don't need to worry about it for that case. (Well, not that we could even get copy elision for this example, since `*this` is a glvalue, but it's good to have this information for reference.) Named Return Value Optimization (NRVO) is technically an optional feature which compilers do not necessarily need to implement, but according to cppreference, this optimization would, in a conformant compiler, also ignore const qualifications.

It should be noted, however, that NRVO is explicitly forbidden in constant-evaluated contexts. For the sake of optimization, this isn't really a big deal, since the result is found at compile time, but it may lead to unexpected results. All of the information which I presented here can be found at https://en.cppreference.com/w/cpp/language/copy_elision.

Reply:
It'll still allow copy elision, but it _won't_ allow the copy -> move construction optimization (see "Automatic move from local variables and parameters", in [return statement][])

> The _expression_ is _move-eligible_ if it is a (possibly parenthesized) id-expression that names a variable whose type is:
>
> * a non-volatile object type
> * or a non-volatile rvalue reference to object type
>
> and that variable is declared
>
> * in the body or
> * as a parameter of
>
> the innermost enclosing function or lambda expression.
>
> If the expression is move-eligible, it is treated as an xvalue (thus overload resolution may select the move constructor.

Note that this doesn't say anything about removing const qualification.

[return statement]: https://en.cppreference.com/w/cpp/language/return#Notes